pkg/rbd/rbd_journal.go

```diff
@@ -0,0 +1,307 @@
+/*
+Copyright 2019 The Ceph-CSI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rbd
+
+import (
+	"fmt"
+
+	"github.com/ceph/ceph-csi/pkg/util"
+
+	"github.com/pkg/errors"
+	"k8s.io/klog"
+)
+
+func validateNonEmptyField(field, fieldName, structName string) error {
+	if field == "" {
+		return fmt.Errorf("value '%s' in '%s' structure cannot be empty", fieldName, structName)
+	}
+
+	return nil
+}
+
+func validateRbdSnap(rbdSnap *rbdSnapshot) error {
+	var err error
+
+	if err = validateNonEmptyField(rbdSnap.RequestName, "RequestName", "rbdSnapshot"); err != nil {
+		return err
+	}
+
+	if err = validateNonEmptyField(rbdSnap.Monitors, "Monitors", "rbdSnapshot"); err != nil {
+		return err
+	}
+
+	if err = validateNonEmptyField(rbdSnap.AdminID, "AdminID", "rbdSnapshot"); err != nil {
+		return err
+	}
+
+	if err = validateNonEmptyField(rbdSnap.Pool, "Pool", "rbdSnapshot"); err != nil {
+		return err
+	}
+
+	if err = validateNonEmptyField(rbdSnap.RbdImageName, "RbdImageName", "rbdSnapshot"); err != nil {
+		return err
+	}
+
+	if err = validateNonEmptyField(rbdSnap.ClusterID, "ClusterID", "rbdSnapshot"); err != nil {
+		return err
+	}
+
+	return err
+}
+
+func validateRbdVol(rbdVol *rbdVolume) error {
+	var err error
+
+	if err = validateNonEmptyField(rbdVol.RequestName, "RequestName", "rbdVolume"); err != nil {
```

Comment:
Cant we do this in a single function with a generic input param?

Reply:
I assume you are talking about a single function that takes in rbdVol/snapVol as the input and uses reflection and then checks the fields rather than the 2 functions.

I did attempt this when @Madhu-1 pointed out a similar reflection [comment](https://github.com/ceph/ceph-csi/pull/381#discussion_r286442146), but it looked more tedious than the current, and I was not able to do cleanly as well.

@humblec Do you have examples that I can look at to get this done?

Reply:
@ShyamsundarR my thought was that, just pass the struct and do rest from the `validateNonEmptyField()`

Reply:
FWIW this pattern is from the FS plugin part of the [code](https://github.com/ceph/ceph-csi/blob/7599d42f1bf03e0d605d7cf78b5150edcb2cc68e/pkg/cephfs/volumeoptions.go#L35-L73).

`validateNonEmptyField` is defined to validate a single field, and `validateRbdVol` is defined to validate certain elements of the RbdVolume structure (similarly `validateRbdSnap`).

Just passing the struct to `validateNonEmptyField` is not sufficient as we need to tell the function which field to validate as well.

Here is what I posted in the other comment (see code snippet [here](https://play.golang.org/p/jzhpeSPs1OW)).

In the posted snippet we could possibly also remove the first `string` argument named `field`, and use the `fieldName` argument to find the element in the structure and then compare and throw the required error. All this requires reflection as stated. @humblec **Is this what you are looking for?**

My next question is, **Is this essential for this patch, or can it be done later? As it impacts no functionality afaict.**

In an ideal world I would probably have a single `validateStruct` (or similarly named) function that takes in an `interface{}` argument, and using reflection determines which fields need validation based on the passed in type of the structure. Thus, merging the 2 validate functions here into a single function.

Reply:
> In an ideal world I would probably have a single `validateStruct` (or similarly named) function that takes in an `interface{}` argument, and using reflection determines which fields need validation based on the passed in type of the structure. Thus, merging the 2 validate functions here into a single function.

[Here](https://play.golang.org/p/RZRq-N-xFA4) is a version of the above, let me know if this works better for you.

The intention being, removing the `validateRbdVol` and `validateRbdSnap` functions and replacing it with a single `validateRbdStruct` that will reflect on the structure type and validate appropriate fields.

@Madhu-1 PTAL as well, to cover the other comment posted [here](https://github.com/ceph/ceph-csi/pull/381#discussion_r286442146) on using reflection.

Reply:
@ShyamsundarR lets look into the pattern.

```
func validateNonEmptyField(field, fieldName, structName string) error {
	if field == "" {
		return fmt.Errorf("value '%s' in '%s' structure cannot be empty", fieldName, structName)
	}

	return nil
}

func validateRbdSnap(rbdSnap *rbdSnapshot) error {
	var err error

	if err = validateNonEmptyField(rbdSnap.RequestName, "RequestName", "rbdSnapshot"); err != nil {
		return err
	}

	if err = validateNonEmptyField(rbdSnap.Monitors, "Monitors", "rbdSnapshot"); err != nil {
		return err
	}


```
 First of all `validateNonEmptyField()` **does NOT even use** the second and third field other than logging!! That does not make any sense to me.  

Secondly, the `validateNonEmptyField()` is "JUST" validating whether first arg is **`""`**. If thats the case why cant we do that in `validateRbdSnap()` or `validateRbdVol()`. ?

Also:
- first and second parameters are **no** different in its "name" or "string" 
- Third arg is same for a struct

I dont feel this validation and repetition correct , so the comment.

>My next question is, Is this essential for this patch, or can it be done later? As it impacts no functionality afaict.
>

I feel this has to be improved/corrected regardless of the functional impact.

Reply:
> @ShyamsundarR lets look into the pattern.
> 
> ```
> func validateNonEmptyField(field, fieldName, structName string) error {
> 	if field == "" {
> 		return fmt.Errorf("value '%s' in '%s' structure cannot be empty", fieldName, structName)
> 	}
> 
> 	return nil
> }
> 
> func validateRbdSnap(rbdSnap *rbdSnapshot) error {
> 	var err error
> 
> 	if err = validateNonEmptyField(rbdSnap.RequestName, "RequestName", "rbdSnapshot"); err != nil {
> 		return err
> 	}
> 
> 	if err = validateNonEmptyField(rbdSnap.Monitors, "Monitors", "rbdSnapshot"); err != nil {
> 		return err
> 	}
> ```
> 
> First of all `validateNonEmptyField()` **does NOT even use** the second and third field other than logging!! That does not make any sense to me.
> 
> Secondly, the `validateNonEmptyField()` is "JUST" validating whether first arg is **`""`**. If thats the case why cant we do that in `validateRbdSnap()` or `validateRbdVol()`. ?
> 
> Also:
> 
>     * first and second parameters are **no** different in its "name" or "string"
> 
>     * Third arg is same for a struct
> 
> 
> I dont feel this validation and repetition correct , so the comment.

Did you see the pattern that I pointed to in CephFS plugin code that uses the same exact method?

Not saying it is the best or correct, just saying that the code is as it exists in other places, and is borrowed from **this** code base.

> 
> > My next question is, Is this essential for this patch, or can it be done later? As it impacts no functionality afaict.
> 
> I feel this has to be improved/corrected regardless of the functional impact.

I agree, the question is does that have to be this patch as it is an existing pattern in the code?

Also, have you see the potential use of reflection as suggested by @Madhu-1 and my thoughts on using it as posted [here](https://github.com/ceph/ceph-csi/pull/381#discussion_r288145690)? If so what are your thoughts on the same?

Reply:
@ShyamsundarR we can refactor the code later for this function ( let me come up with a better approach if possible)

Reply:
>Did you see the pattern that I pointed to in CephFS plugin code that uses the same exact method?

Not saying it is the best or correct, just saying that the code is as it exists in other places, and is borrowed from this code base.
>

Well, I understand that :) . At the same time, it does not mean we have to wait and do the correction later. 
>I agree, the question is does that have to be this patch as it is an existing pattern in the code?
> 

This is a tempting question :) , considering the release readiness and the number of PRs waiting for this to land, I think I have to rethink .. Let me do one more review before commenting further.